Readme.md
# ⚠️ Unmaintained ⚠️

This repository is unmaintained, and this [functionality has been added to superagent core](https://visionmedia.github.io/superagent/#retrying-requests).

# superagent-retry

  Extends the node version of [`visionmedia/superagent`][superagent]'s `Request`, adds a `.retry` method to add retrying logic to the request. Calling this will retry the request however many additional times you'd like.


  [superagent]: https://github.com/visionmedia/superagent

## Usage

```javascript
var superagent = require('superagent');
require('superagent-retry')(superagent);

superagent
  .get('https://segment.io')
  .retry(2) // retry twice before responding
  .end(onresponse);


function onresponse (err, res) {
  console.log(res.status, res.headers);
  console.log(res.body);
}
```

You can optionally add a delay to each retry.

```javascript
superagent
  .get('https://segment.io')
  .retry(2, 500) // retry twice with 0.5 second delay
  .end(onresponse);
```

## Retrying Cases

  Currently the retrying logic checks for:

  * ECONNRESET
  * ECONNREFUSED
  * ETIMEDOUT
  * EADDRINFO
  * ESOCKETTIMEDOUT
  * superagent client timeouts
  * bad gateway errors (502, 503, 504 statuses)
  * Internal Server Error (500 status)


## License

(The MIT License)

Copyright (c) 2013 Segmentio &lt;friends@segment.io&gt;

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.md

0.7.0 / 2016-12-15
==================

  * [FEATURE] Retry with optional delay

0.6.0 / 2016-05-16
==================

  * [FEATURE] Retry on 500 errors

0.5.1 / 2014-10-08
==================

 * retain request path
 * downgrade superagent for backfilling fix

0.5.0 / 2014-10-03
==================

 * upgrade superagent to 0.20

0.4.0 / 2014-09-30
==================

 * update superagent

0.3.0 / 2014-04-30
==================

 * update superagent

0.2.0 / 2014-03-06
==================

 * re-set headers
 * update superagent

0.1.1 / 2013-10-23
==================

  * Added support for 502, 504 responses ([justincy](https://github.com/justincy))

0.1.0 / 2013-09-03
==================

  * Initial Version


test/test.js
var express     = require('express')
  , agent       = require('superagent')
  , should      = require('should')
  , assert      = require('assert')
  , http        = require('http');


http.globalAgent.maxSockets = 2000;


require('../')(agent);


describe('superagent-retry', function () {

  describe('errors', function () {
    var requests = 0
      , port = 10410
      , success
      , app = express()
      , server;

    before(function (done) {
      app.get('/', function (req, res, next) {
        requests++;
        if (requests < 4) res.send(503);
        else res.send('hello!');
      });

      app.get('/user-error', function (req, res, next) {
        requests++;
        if (requests < 4) res.send(500);
        else res.send('hello!');
      });

      app.get('/user-error-with-delay', function (req, res, next) {
        setTimeout(function () {
          success = true;
        }, 500);
        return res.send(success ? 200 : 500);
      });

      server = app.listen(port, done);
    });

    beforeEach(function () {
      requests = 0
    });

    it('should retry on errors', function (done) {

      agent
        .get('http://localhost:' + port)
        .end(function (err, res) {
          res.status.should.eql(503);
        });

      setTimeout(function () {
        agent
          .get('http://localhost:' + port)
          .retry(5)
          .end(function (err, res) {
            res.text.should.eql('hello!');
            requests.should.eql(4);
            done(err);
          });
      }, 100);
    });

    it('should retry on user defined errors', function (done) {

      agent
        .get('http://localhost:' + port + '/user-error')
        .end(function (err, res) {
          res.status.should.eql(500);
        });

      setTimeout(function () {
        agent
          .get('http://localhost:' + port + '/user-error')
          .retry(5, function (err, res) {
            return res && res.status === 500;
          })
          .end(function (err, res) {
            res.text.should.eql('hello!');
            requests.should.eql(4);
            done(err);
          });
      }, 100);
    });

    it('should wait before retry', function (done) {

      agent
        .get('http://localhost:' + port + '/user-error-with-delay')
        .end(function (err, res) {
          res.status.should.eql(500);
        });

      setTimeout(function () {
        agent
          .get('http://localhost:' + port + '/user-error-with-delay')
          .retry(5, function (err, res) {
            return res && res.status === 500;
          })
          .delayRetry(600)
          .end(function (err, res) {
            res.status.should.eql(200);
            done(err);
          });
      }, 100);
    });

    after(function (done) { server.close(done); });
  });

  describe('500 errors', function () {
    var requests = 0
      , port = 10410
      , app = express()
      , server;

    before(function (done) {
      app.get('/', function (req, res, next) {
        requests++;
        if (requests < 4) res.send(500);
        else res.send('hello!');
      });

      server = app.listen(port, done);
    });

    it('should retry on errors', function (done) {

      agent
        .get('http://localhost:' + port)
        .end(function (err, res) {
          res.status.should.eql(500);
        });

      setTimeout(function () {
        agent
          .get('http://localhost:' + port)
          .retry(5)
          .end(function (err, res) {
            console.log('requests', requests)
            res.text.should.eql('hello!');
            requests.should.eql(4);
            done(err);
          });
      }, 100);
    });

    after(function (done) { server.close(done); });
  });

  describe('resets', function () {
    var requests = 0
      , port = 10410
      , app = express()
      , server;

    before(function (done) {
      server = app.listen(port, done);
    });

    it('should retry client timeouts', function (done) {
      app.get('/client-timeouts', function (req, res, next) {
        requests++;
        if (requests > 10) res.send('hello!');
      });

      var url = 'http://localhost:' + port + '/client-timeouts';

      agent
        .get(url)
        .timeout(10)
        .end(function (err, res) {
          should.exist(err);
        });

      agent
        .get(url)
        .timeout(2)
        .retry(20)
        .end(function (err, res) {
          res.text.should.eql('hello!');
          done();
        });
    });

    it('should retry with the same headers', function(done){
      var url = 'http://localhost:' + port + '/headers';
      var requests = 0;

      app.get('/headers', function(req, res){
        if (++requests > 3) return res.send(req.headers);
      });

      agent
        .get(url)
        .set('Accept', 'application/json')
        .set('X-Foo', 'baz')
        .timeout(10)
        .retry(4)
        .end(function(err, res){
          assert('baz' == res.body['x-foo']);
          assert('application/json' == res.body['accept']);
          done();
        });
    })

    it('should re-send data and headers correctly', function(done){
      var url = 'http://localhost:' + port + '/data';
      var requests = 0;

      app.post('/data', express.bodyParser(), function(req, res){
        if (++requests < 3) return;
        res.send({ body: req.body, headers: req.headers });
      });

      agent
        .post(url)
        .type('json')
        .send({ data: 1 })
        .timeout(10)
        .retry(4)
        .end(function(err, res){
          assert(1 == res.body.body.data);
          assert('application/json' == res.body.headers['content-type']);
          done();
        });
    })

    it('should retry on server resets', function (done) {
      var requests = 0;

      app.get('/server-timeouts', function (req, res, next) {
        requests++;
        if (requests > 10) return res.send('hello!');
        res.setTimeout(1);
      });

      var url = 'http://localhost:' + port + '/server-timeouts';

      agent
        .get(url)
        .end(function (err, res) {
          err.code.should.eql('ECONNRESET');
        });

      agent
        .get(url)
        .retry(20)
        .end(function (err, res) {
          res.text.should.eql('hello!');
          done();
        });
    });

    it('should retry on server connection refused', function (done) {
      var url = 'http://localhost:' + (port+1) + '/hello';
      var request = agent.get(url);
      var allowedRetries = 10;
      var allowedTries = allowedRetries + 1;
      var triesCount = 0

      var oldEnd = request.end;
      request.end = function(fn) {
        triesCount++;
        oldEnd.call(request, fn);
      };

      request
        .retry(allowedRetries)
        .end(function (err, res) {
          err.code.should.eql('ECONNREFUSED');
          triesCount.should.eql(allowedTries);
          done();
        });
    });

    it('should retry with the same querystring', function(done){
      var requests = 0;

      app.get('/qs-data', function(req, res){
        if (++requests > 10) return res.json({ foo: req.query.foo });
        res.setTimeout(1);
      });

      var url = 'http://localhost:' + port + '/qs-data';

      agent
        .get(url)
        .retry(20)
        .query({ foo: 'bar' })
        .end(function(err, res){
          res.body.foo.should.eql('bar');
          done();
        })
    });
  });
})


lib/index.js
var retries = require('./retries');


/**
 * Add to the request prototype.
 */

module.exports = function (superagent) {
  var Request = superagent.Request;
  Request.prototype.retry = retry;
  return superagent;
};


/**
 * Export retries for extending
 */

module.exports.retries = retries;


/**
 * Sets the amount of times to retry the request
 * @param  {Number} count
 * @param  {Number} [delay] optional delay in ms
 */

function retry (retries, delay) {

  var self    = this
    , oldEnd  = this.end;

  retries     = retries === Number(retries) && retries || 1;
  delay       = delay === Number(delay) && delay || 0;

  this.end = function (fn) {
    var timeout = this._timeout;

    function attemptRetry () {
      return oldEnd.call(self, function (err, res) {
        if (!retries-- || !shouldRetry(err, res)) return fn && fn(err, res);

        setTimeout(function() {
          reset(self, timeout);
          attemptRetry();
        }, delay);
      });
    }

    return attemptRetry();
  };

  return this;
}


/**
 * HACK: Resets the internal state of a request.
 */

function reset (request, timeout) {
  var headers = request.req._headers;
  var path = request.req.path;

  request.req.abort();
  request.called = false;
  request.timeout(timeout);
  delete request.req;
  delete request._timer;
  delete request._endCalled;

  for (var k in headers) {
    request.set(k, headers[k]);
  }

  if (!request.qs) {
    request.req.path = path;
  }
}


/**
 * Determine whether we should retry based upon common error conditions
 * @param  {Error}    err
 * @param  {Response} res
 * @return {Boolean}
 */

function shouldRetry (err, res) {
  return retries.some(function (check) { return check(err, res); });
}


lib/retries.js
/**
 * Common retry conditions
 */

module.exports = [
  econnreset,
  econnrefused,
  etimedout,
  eaddrinfo,
  esockettimedout,
  crossDomain,
  gateway,
  timeout,
  internal
];


/**
 * Connection reset detection
 */

function econnreset (err, res) {
  return err && err.code === 'ECONNRESET';
}

/**
 * Connection refused detection
 */

function econnrefused (err, res) {
  return err && err.code === 'ECONNREFUSED';
}

/**
 * Timeout detection
 */

function etimedout (err, res) {
  return err && err.code === 'ETIMEDOUT';
}


/**
 * Can't get address info
 */

function eaddrinfo (err, res) {
  return err && err.code === 'EADDRINFO';
}


/**
 * Socket timeout detection
 */

function esockettimedout (err, res) {
  return err && err.code === 'ESOCKETTIMEDOUT';
}

/**
 * Superagent crossDomain error
 */

function crossDomain (err, res) {
  return err && err.crossDomain;
}

/**
 * Internal server error
 */

function internal (err, res) {
  return res && res.status === 500;
}

/**
 * Bad gateway error detection
 */

function gateway (err, res) {
  return res && [502,503,504].indexOf(res.status) !== -1;
}


/**
 * Superagent timeout errors
 */

function timeout (err, res) {
  return err && /^timeout of \d+ms exceeded$/.test(err.message);
}
